File: backend/app/schemas/investigation.py
from datetime import datetime, timedelta
from uuid import UUID

from pydantic import BaseModel

from app.schemas.alert import (
    AlertInvestigationHost,
    AlertInvestigationTimeline,
    AlertResponse,
)
from app.schemas.event import EventResponse


class WorkspaceAnchor(BaseModel):
    type: str
    id: UUID


class WorkspaceOffenseSummary(BaseModel):
    id: UUID
    offense_number: int
    host_id: UUID
    host_name: str | None = None
    title: str
    description: str | None = None
    risk_level: str
    status: str
    event_count: int
    alert_count: int
    incident_id: UUID | None = None
    timeline: list = []
    related_users: list = []
    alerts: list[dict] = []
    events: list[dict] = []


class WorkspaceIncidentSummary(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    severity: str
    status: str
    host_id: UUID | None = None
    created_at: datetime
    resolved_at: datetime | None = None
    notes: list[dict] = []
    alert_ids: list[str] = []


class InvestigationWorkspaceResponse(BaseModel):
    anchor: WorkspaceAnchor
    alert: AlertResponse | None = None
    offense: WorkspaceOffenseSummary | None = None
    incident: WorkspaceIncidentSummary | None = None
    host: AlertInvestigationHost | None = None
    events: list[EventResponse] = []
    timelines: list[AlertInvestigationTimeline] = []
    linked_alerts: list[AlertResponse] = []

File: frontend/lib/types/investigation.ts
export interface WorkspaceAnchor {
  type: "alert" | "offense" | "incident";
  id: string;
}

export interface WorkspaceHost {
  id: string;
  name: string;
  hostname: string | null;
  status: string;
  ip_address: string | null;
  risk_score: number | null;
}

export interface WorkspaceAlert {
  id: string;
  host_id: string;
  title: string;
  description: string | null;
  severity: string;
  status: string;
  confidence: number | null;
  mitre_technique_id: string | null;
  created_at: string;
}

export interface WorkspaceOffense {
  id: string;
  offense_number: number;
  host_id: string;
  host_name: string | null;
  title: string;
  description: string | null;
  risk_level: string;
  status: string;
  event_count: number;
  alert_count: number;
  incident_id: string | null;
  timeline: Array<Record<string, unknown>>;
  related_users: string[];
  alerts: Array<{ id: string; title: string; severity: string; status: string; created_at: string }>;
  events: Array<{ id: string; event_type: string; description: string | null; severity: string; timestamp: string }>;
}

export interface WorkspaceIncident {
  id: string;
  title: string;
  description: string | null;
  severity: string;
  status: string;
  host_id: string | null;
  created_at: string;
  resolved_at: string | null;
  notes: Array<{ id: string; content: string; user_id: string; created_at: string }>;
  alert_ids: string[];
}

export interface WorkspaceEvent {
  id: string;
  event_type: string;
  severity: string;
  description: string | null;
  timestamp: string;
}

export interface WorkspaceTimeline {
  id: string;
  title: string;
  severity: string;
  confidence: number;
  started_at: string;
  status: string;
}

export interface InvestigationWorkspace {
  anchor: WorkspaceAnchor;
  alert: WorkspaceAlert | null;
  offense: WorkspaceOffense | null;
  incident: WorkspaceIncident | null;
  host: WorkspaceHost | null;
  events: WorkspaceEvent[];
  timelines: WorkspaceTimeline[];
  linked_alerts: WorkspaceAlert[];
}
